sync/server.md

```diff
@@ -0,0 +1,403 @@
+The server will use the standard module mechanism, `npm`, for installing and managing these packages.
+Npm `scripts` will be used extensively for running CI and local development tasks, executing cli tools directly e.g. `nodemon`, `standard`, `mocha`.
+Build tools such as `grunt` or `gulp` will *not* be used as they introduce extra overhead where a simple cli command will suffice. See https://www.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/ for more info.
+
+## Architecture
+
+The Sync server is intended to be horizontally scalable.
+As such, it will us a [shared-nothing-architecture](https://en.wikipedia.org/wiki/Shared-nothing_architecture).
+In addition, the server will be developed following the [12 factor app](https://12factor.net/) pattern.
+The architecture will be kept simple, where appropriate.
+There will be a single server process which serves as the following:
+
+* Static file server for the Admin UI
+* API for managing the Data Schema and other GraphQL & Data Sync features
+* GraphQL API for Apps to send Queries & Mutations to
+* Subscriptions endpoint for Apps to connect to
+
+TODO: Updated architecture diagram based on https://docs.google.com/drawings/d/1OmMY1bbI8HDhlZ_cIirv7FYTEqRRV-Hh-g0ANtYEgBY/edit
+
+The main developer flow when using the Sync Server will be something like this:
+
+* Open the Data Sync Admin UI
+* Define the Data Schema
+* Configure Data Source(s)
+* Configure Resolvers for any Queries or Mutations
+* Update/modify the Data Schema as needed, continuously saving and trying out the Schema
+
+The last part of the flow is very important.
+This implies a fast feedback loop when developing your Data Schema.
+To ensure the feedback loop is as fast as possible, the architecture of sync will be slightly different that the proof of concept architecture.
+Both the backend for the Admin UI, and the GraphQL API that Apps talk to will be part of the 1 Data Sync server.
+This server will continously run, without any restart needed when the Data Schema changes.
+To facilitate this, the mounting of the graphql express middleware will need to be done in such a way that it always rebuilds the schema based on the latests available. In reality, the parsed schema should be cached, and only reloaded/reparsed if it has changed. Here is a sample graphql endpoint mounting in express:
+
+```
+const tracing = true
+app.use('/graphql', bodyParser.json(), function (req, res, next) {
+  const schema = require('./lib/schemaParser').parseFromFile(SCHEMA_FILE, DATA_SOURCES_FILE, RESOLVER_MAPPINGS_FILE)
+  const graphql = graphqlExpress({ schema, tracing })
+  graphql(req, res, next)
+})
+```
+
+As the server needs to store resources, such as the Data Schema, Resolver Mappings and Webhooks, a data store is required.
+The Data Sync server is intended to run in a Kubernetes environment such as Openshift.
+This allows the Data Sync server to use a ServiceAccount token to save these resources in 1 or more Secrets.
```

Comment:
After some discussion with @darahayes, this is unlikely to be the way this works.
Up for discussion, but at the moment I'm thinking of shipping postgres with the data sync server. See comment below for more context https://github.com/aerogear/proposals/pull/40/files#r198116098